File: 518.py
// Time Complexity : O(n * k) where k is length of coins and n is amount
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        
        for coin in coins:
            for x in range(coin, amount + 1):
                dp[x] += dp[x - coin]
        return dp[amount]

File: 518.java
class Solution {
    public int change(int amount, int[] coins) {
        
        return dp_coin_imp(amount, coins);
    }
    
    public int dp_coin_imp(int amount, int[] coins) {
        
        int[] dp = new int[amount + 1];
        dp[0] = 1;
        
        for (int i = 0; i < coins.length; i++) {
            int coin = coins[i];
            for (int j = coin; j < amount + 1; j++) {
                dp[j] += dp[j - coin];
            }
        }
        return dp[amount];
    }
}
//Runtime: O(coins.length * amount)
//Space: O(amount)